From: Tiger Yang <tiger.yang@oracle.com>
Date: Fri, 6 Mar 2009 10:19:30 +0800
Subject: [PATCH] ocfs2: reserve xattr block for new directory with inline data
Patch-mainline: 2.6.29

If this is a new directory with inline data, we choose to
reserve the entire inline area for directory contents and
force an external xattr block.

Signed-off-by: Tiger Yang <tiger.yang@oracle.com>
Acked-by: Joel Becker <joel.becker@oracle.com>
Signed-off-by: Mark Fasheh <mfasheh@suse.com>
---
 fs/ocfs2/xattr.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

Index: linux-2.6.27-sle11_ocfs2_update2/fs/ocfs2/xattr.c
===================================================================
--- linux-2.6.27-sle11_ocfs2_update2.orig/fs/ocfs2/xattr.c
+++ linux-2.6.27-sle11_ocfs2_update2/fs/ocfs2/xattr.c
@@ -450,8 +450,12 @@ int ocfs2_calc_xattr_init(struct inode *
 	 * when blocksize = 512, may reserve one more cluser for
 	 * xattr bucket, otherwise reserve one metadata block
 	 * for them is ok.
+	 * If this is a new directory with inline data,
+	 * we choose to reserve the entire inline area for
+	 * directory contents and force an external xattr block.
 	 */
 	if (dir->i_sb->s_blocksize == OCFS2_MIN_BLOCKSIZE ||
+	    (S_ISDIR(mode) && ocfs2_supports_inline_data(osb)) ||
 	    (s_size + a_size) > OCFS2_XATTR_FREE_IN_IBODY) {
 		ret = ocfs2_reserve_new_metadata_blocks(osb, 1, xattr_ac);
 		if (ret) {
